sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,270 @@
+        Export(buffer_copy, was_force_flush_called);
+        auto end = std::chrono::steady_clock::now(); 
+        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+
+        timeout = schedule_delay_millis_ - duration;
+    }
+}
+
+void BatchSpanProcessor::Export(std::unique_ptr<common::CircularBuffer<Recordable>>& buffer,
+                                const bool was_force_flush_called)
+{
+    std::vector<std::unique_ptr<Recordable>> spans_arr;
+    
+    buffer->Consume(
+    buffer->size(), [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+        range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+        std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+        ptr.Swap(swap_ptr);
+        spans_arr.push_back(std::unique_ptr<Recordable>(swap_ptr.release()));
+        return true;
+        });
+    });
+
+    exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));    
+
+    if(is_shutdown_ == true) return;
+
+    // Notify the main thread in case this export was the result of a force flush.
+    if(was_force_flush_called == true)
+    {
+        is_force_flush_ = false;
+        is_force_flush_notified_ = true;
+
+        while(is_force_flush_notified_ == true)
+        {
+            force_flush_cv_.notify_one();
+        }
+    }
+}
+
+std::unique_ptr<common::CircularBuffer<Recordable>> BatchSpanProcessor::CopySpans(
+    const bool was_force_flush_called
+)
+{
+    // Make a copy of the buffer to avoid blocking other producers.
+    std::unique_ptr<common::CircularBuffer<Recordable>> buffer_copy(
+        new common::CircularBuffer<Recordable>(max_queue_size_)
+    );
+
+    if(was_force_flush_called == true)
+    {
+        buffer_.swap(buffer_copy);
+    }
+    else
+    {
+        // Get the appropriate size
+        const size_t num_spans_to_export = buffer_->size() >= max_export_batch_size_ ? 
+                                                max_export_batch_size_ : buffer_->size();
+
+        buffer_->Consume(
+            num_spans_to_export, [&](CircularBufferRange<AtomicUniquePtr<Recordable>> range) noexcept {
+                range.ForEach([&](AtomicUniquePtr<Recordable> &ptr) {
+                std::unique_ptr<Recordable> swap_ptr = std::unique_ptr<Recordable>(nullptr);
+                ptr.Swap(swap_ptr);
+                buffer_copy->Add(swap_ptr);
+                return true;
+                });
+            });
+    }
+
+    return buffer_copy;
+}
+
+void BatchSpanProcessor::DrainQueue()
+{
+    while(buffer_->empty() == false) Export(buffer_, false);
+}
+
+void BatchSpanProcessor::Shutdown(std::chrono::microseconds timeout) noexcept 
+{
+    auto start_time = std::chrono::steady_clock::now();
+
+    is_shutdown_ = true;
+
+    // Notify thread, drain the queue and wait for the thread.
+    while(is_shutdown_ == true) cv_.notify_one();
```

Comment:
This might need `volatile` to be added to `is_shutdown_` on over-optimizing compilers?

Reply:
Right.

What do you think about using `std::atomic<bool>`?

Reply:
Changed to volatile 👍